### PR TITLE
Add support for deployment protection rules

### PR DESCRIPTION
**Pull request by abhishekbiyala**
Fixes https://github.com/google/go-github/issues/3022

### CODE REVIEW

File: github/repos_deployment_protection_rules.go [hunk 1]
```diff
@@ -0,0 +1,147 @@
+// Copyright 2023 The go-github AUTHORS. All rights reserved.
```
Comment (gmlewis) [inline github/repos_deployment_protection_rules.go:1]:
```suggestion
// Copyright 2024 The go-github AUTHORS. All rights reserved.
```

File: github/repos_deployment_protection_rules.go [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
+// license that can be found in the LICENSE file.
+
+package github
+
+import (
+	"context"
+	"fmt"
+)
+
+type CustomDeploymentProtectionRuleApp struct {
```
Comment (gmlewis) [inline github/repos_deployment_protection_rules.go:13]: Please add a descriptive godoc-style comment for this exported struct.

File: github/repos_deployment_protection_rules.go [hunk 3]
```diff
@@ -0,0 +1,147 @@
...
+
+// CustomDeploymentProtectionRule represents a single deployment protection rule for an environment.
+type CustomDeploymentProtectionRule struct {
+	ID      *int64                             `json:"id,omitempty"`
+	NodeID  *string                            `json:"node_id,omitempty"`
+	Enabled *bool                              `json:"enabled,omitempty"`
+	App     *CustomDeploymentProtectionRuleApp `json:"app,omitempty"`
+}
+
+// ListDeploymentProtectionRuleResponse represents the format of response that comes back when you list deployment protection rules.
```
Comment (gmlewis) [inline github/repos_deployment_protection_rules.go:28]:
```suggestion
// ListDeploymentProtectionRuleResponse represents the response that comes back when you list deployment protection rules.
```

File: github/repos_deployment_protection_rules.go [hunk 4]
```diff
@@ -0,0 +1,147 @@
...
+type ListDeploymentProtectionRuleResponse struct {
+	TotalCount      *int                              `json:"total_count,omitempty"`
+	ProtectionRules []*CustomDeploymentProtectionRule `json:"custom_deployment_protection_rules,omitempty"`
+}
+
+// ListCustomDeploymentRuleIntegrationsResponse represents the slightly different format of response that comes back when you list custom deployment rule integrations.
```
Comment (gmlewis) [inline github/repos_deployment_protection_rules.go:34]:
```suggestion
// ListCustomDeploymentRuleIntegrationsResponse represents the slightly different response that comes back when you list custom deployment rule integrations.
```

File: github/repos_deployment_protection_rules.go [hunk 5]
```diff
@@ -0,0 +1,147 @@
...
+type ListCustomDeploymentRuleIntegrationsResponse struct {
+	TotalCount *int                                 `json:"total_count,omitempty"`
+	Apps       []*CustomDeploymentProtectionRuleApp `json:"available_custom_deployment_protection_rule_integrations,omitempty"`
```
Comment (gmlewis) [inline github/repos_deployment_protection_rules.go:37]: How about "Apps" => "AvailableIntegrations"?

File: github/repos_deployment_protection_rules.go [hunk 6]
```diff
@@ -0,0 +1,147 @@
...
+}
+
+// CustomDeploymentProtectionRuleRequest represents a deployment protection rule request.
+type CustomDeploymentProtectionRuleRequest struct {
+	IntegrationID *int64 `json:"integration_id,omitempty"`
+}
+
+// GetAllDeploymentProtectionRules get all the deployment protection rules for an environment.
```
Comment (gmlewis) [inline github/repos_deployment_protection_rules.go:45]:
```suggestion
// GetAllDeploymentProtectionRules gets all the deployment protection rules for an environment.
```

**Review by gmlewis — CHANGES_REQUESTED** [review]
Thank you, @abhishekbiyala !
Just a few minor tweaks, please, then we will be ready for a second LGTM+Approval from any other contributor to this repo before merging.

**Review by gmlewis — APPROVED** [review]
Thank you, @abhishekbiyala !
LGTM.

Awaiting second LGTM+Approval from any other contributor to this repo before merging.

**Comment by fchimpan** [comment]
Hi, @abhishekbiyala 
Thanks your implementation. LGTM!

**Comment by gmlewis** [comment]
Thank you, @fchimpan !
Merging.

**Pull request merged**